src/main/java/hudson/plugins/analysis/core/HealthAwarePublisher.java

```diff
@@ -210,8 +220,33 @@ protected void updateBuildResult(final BuildResult result, final PluginLogger lo
      *             a better error message, if it can do so, so that users have
      *             better understanding on why it failed.
      */
-    protected abstract BuildResult perform(AbstractBuild<?, ?> build, PluginLogger logger)
-            throws InterruptedException, IOException;
+    @Deprecated
+    protected BuildResult perform(AbstractBuild<?, ?> build, PluginLogger logger)
+            throws InterruptedException, IOException {
+        return perform((Run) build, null, logger);
+    }
+
+    protected /*abstract*/ BuildResult perform(Run<?, ?> run, FilePath workspace, PluginLogger logger) throws InterruptedException, IOException{
+        if (isOverridden(HealthAwarePublisher.class, getClass(),
+                "perform", AbstractBuild.class, PluginLogger.class) && run instanceof AbstractBuild) {
+            return perform((AbstractBuild) run, logger);
+        } else {
+            // Runtime error to force overriding this method
+            throw new AbstractMethodError("you must override the new overload of perform(Run, FilePath, Launcher, PluginLogger)");
+        }
+    }
+
+    /**
+     * Util.isOverriden does not work on non-public methods, that's why this one is used here.
+     */
+    private static boolean isOverridden(@Nonnull Class base, @Nonnull Class derived, @Nonnull String methodName, @Nonnull Class... types) {
+        try {
+            return !base.getDeclaredMethod(methodName, types).equals(
+                    derived.getDeclaredMethod(methodName,types));
```

Comment:
This is not correct. `getDeclaredMethod` ignores inheritance, so it will _normally_ throw `NoSuchMethodException` if the method is not overridden.

Anyway even if you fixed that, by treating NSME as `false`, a single check like this cannot work. Consider

```java
public abstract class Base {
    protected abstract void method();
}
public abstract class Intermediate extends Base {
    protected void method() {}
}
public class Derived extends Intermediate {}
```

You expect `isOverridden` to return true here despite the NSME from `derived.getDeclaredMethod`. The only to fix it is to use a loop via `Class.getSuperclass`, effectlvely duplicating the algorithm in `Class.getMethod` without the restriction to `public` methods.